### ISSUE TITLE
When using a reloadable deployment, the build after entering continuous build reports the wrong total build time.

### ISSUE THREAD

**Issue by ymwang1984** [gradle/gradle#6789]
This issue has been extensively discussed in [this forum question](https://discuss.gradle.org/t/why-does-plays-runplaybinary-always-load-twice/28158).

### Current Behavior

When we start the Play application, we expect it to load just once before starting to listen for requests. However, currently it loads twice.

Per @bigdaz from Gradle Inc in the [forum question](https://discuss.gradle.org/t/why-does-plays-runplaybinary-always-load-twice/28158/3)

> The fact that we execute the build twice on initial startup is a bug.

### Context

Server start up becomes slower, which impacts developer productivity.

### Steps to Reproduce (for bugs)

This can be easily reproduced in sample Play apps that are inside Gradle.

```
[info] p.c.s.AkkaHttpServer - Listening for HTTP on /0:0:0:0:0:0:0:0:9000
 
> Task :runPlayBinary
Running Play App (:runPlayBinary) at http://localhost:9000/
 
BUILD SUCCESSFUL in 54s
7 actionable tasks: 7 executed
 
Reloadable deployment detected. Entering continuous build.
 
BUILD SUCCESSFUL in 54s
7 actionable tasks: 1 executed, 6 up-to-date
 
Waiting for changes to input files of tasks... (ctrl-d to exit)
```

### Your Environment

 * Build scan URL: 
It happens both on Mac and Linux.

**Comment by big-guy** [comment]
I think the bug here might be with the output.

If you watch the console output, is there a ~minute between the first BUILD SUCCESSFUL and the second? Locally, I see the same behavior, but the second time is definitely wrong and the build is completely up-to-date. It looks like the time isn't reset for the second build.

**Comment by ymwang1984** [comment]
Thanks @big-guy 

This is the info log. We ran on play-2.4 sample project with Gradle 4.9.

```
ymwang play-2.4 $ ./gradlew runPlayBinary --info
Initialized native services in: /home/ymwang/.gradle/native
The client will now receive all logging from the daemon (pid: 36046). The daemon log file: /home/ymwang/.gradle/daemon/4.9/daemon-36046.out.log
Starting 4th build in daemon [uptime: 1 mins 24.739 secs, performance: 100%, no major garbage collections]
Using 16 worker leases.
Starting Build
Settings evaluated using settings file '/home/ymwang/Downloads/play-2.4/settings.gradle'.
Projects loaded. Root project using build file '/home/ymwang/Downloads/play-2.4/build.gradle'.
Included projects: [root project 'play-2.4']

> Configure project :
Evaluating root project 'play-2.4' using build file '/home/ymwang/Downloads/play-2.4/build.gradle'.
All projects evaluated.
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
file or directory '/home/ymwang/Downloads/play-2.4/app/assets', not found
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
Selected primary task 'runPlayBinary' from project :
Tasks to be executed: [task ':createPlayBinaryAssetsJar', task ':compilePlayBinaryPlayRoutes', task ':compilePlayBinaryPlayTwirlTemplates', task ':compilePlayBinaryScala', task ':processPlayBinaryPlayResources', task ':createPlayBinaryJar', task ':playBinary', task ':runPlayBinary']
:createPlayBinaryAssetsJar (Thread[Task worker for ':',5,main]) started.

> Task :createPlayBinaryAssetsJar
Task ':createPlayBinaryAssetsJar' is not up-to-date because:
  Output property 'archivePath' file /home/ymwang/Downloads/play-2.4/build/playBinary/lib/play-2.4-assets.jar has been removed.
:createPlayBinaryAssetsJar (Thread[Task worker for ':',5,main]) completed. Took 0.012 secs.
:compilePlayBinaryPlayRoutes (Thread[Task worker for ':',5,main]) started.
This JVM does not support getting OS memory, so no OS memory status updates will be broadcast
Initialized native services in: /home/ymwang/.gradle/native

> Task :compilePlayBinaryPlayRoutes
Task ':compilePlayBinaryPlayRoutes' is not up-to-date because:
  Output property 'outputDirectory' file /home/ymwang/Downloads/play-2.4/build/src/play/binary/routesScalaSources has been removed.
  Output property 'outputDirectory' file /home/ymwang/Downloads/play-2.4/build/src/play/binary/routesScalaSources/controllers has been removed.
  Output property 'outputDirectory' file /home/ymwang/Downloads/play-2.4/build/src/play/binary/routesScalaSources/controllers/javascript has been removed.
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
Starting process 'Gradle Worker Daemon 5'. Working directory: /home/ymwang/.gradle/workers Command: /export/apps/jdk/JDK-1_8_0_121/bin/java -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/ymwang/.gradle/caches/4.9/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Worker Daemon 5'
Successfully started process 'Gradle Worker Daemon 5'
Started Gradle worker daemon (0.353 secs) with fork options DaemonForkOptions{executable=/export/apps/jdk/JDK-1_8_0_121/bin/java, minHeapSize=null, maxHeapSize=null, jvmArgs=[], classpath=[/home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.play/routes-compiler_2.10/2.4.0/12b1375ab1c872c0c3b2aa8217881ad8ef8aea88/routes-compiler_2.10-2.4.0.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.play/twirl-api_2.10/1.1.1/b11c77752bfbb3d454a54a3675a5f9ecbe564dba/twirl-api_2.10-1.1.1.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-library/2.10.5/57ac67a6cf6fd591e235c62f8893438e8d10431d/scala-library-2.10.5.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/commons-io/commons-io/2.4/b1b6ea3b7e4aa4f492509a4952029cd8e48019ad/commons-io-2.4.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-lang3/3.4/5fe28b9518e58819180a43a850fbc0dd24b7c050/commons-lang3-3.4.jar], keepAliveMode=SESSION}.
:compilePlayBinaryPlayRoutes (Thread[Task worker for ':',5,main]) completed. Took 0.834 secs.
:compilePlayBinaryPlayTwirlTemplates (Thread[Task worker for ':',5,main]) started.
This JVM does not support getting OS memory, so no OS memory status updates will be broadcast
Initialized native services in: /home/ymwang/.gradle/native

> Task :compilePlayBinaryPlayTwirlTemplates
Task ':compilePlayBinaryPlayTwirlTemplates' is not up-to-date because:
  Output property 'outputDirectory' file /home/ymwang/Downloads/play-2.4/build/src/play/binary/twirlTemplatesScalaSources has been removed.
  Output property 'outputDirectory' file /home/ymwang/Downloads/play-2.4/build/src/play/binary/twirlTemplatesScalaSources/views has been removed.
  Output property 'outputDirectory' file /home/ymwang/Downloads/play-2.4/build/src/play/binary/twirlTemplatesScalaSources/views/html has been removed.
All input files are considered out-of-date for incremental task ':compilePlayBinaryPlayTwirlTemplates'.
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
Starting process 'Gradle Worker Daemon 6'. Working directory: /home/ymwang/.gradle/workers Command: /export/apps/jdk/JDK-1_8_0_121/bin/java -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/ymwang/.gradle/caches/4.9/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Worker Daemon 6'
Successfully started process 'Gradle Worker Daemon 6'
Started Gradle worker daemon (0.395 secs) with fork options DaemonForkOptions{executable=/export/apps/jdk/JDK-1_8_0_121/bin/java, minHeapSize=null, maxHeapSize=null, jvmArgs=[], classpath=[/home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.play/twirl-compiler_2.11/1.1.1/c52e000cfe1895cd1dc9fb5c0d4ac5bbd73713cc/twirl-compiler_2.11-1.1.1.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-compiler/2.11.6/fa7d8a52d53c6571bce334b2c1d3d9546314dde3/scala-compiler-2.11.6.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.play/twirl-api_2.11/1.1.1/b029d9500caec7fe30f9a485fc654ee82d40d404/twirl-api_2.11-1.1.1.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.play/twirl-parser_2.11/1.1.1/8af782249d76c70d411305f94eff19c9906c43af/twirl-parser_2.11-1.1.1.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-reflect/2.11.6/f539c9c9cf312472b3a7771ac85ecda859657a1d/scala-reflect-2.11.6.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang.modules/scala-xml_2.11/1.0.3/8f295fc7620bf5a71eb7e34e71d97207aa534abf/scala-xml_2.11-1.0.3.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang.modules/scala-parser-combinators_2.11/1.0.3/b85f2ff7cbdb5540a4c293a947cc936892040fc9/scala-parser-combinators_2.11-1.0.3.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-library/2.11.6/be3457b4b748df35bffba675d8cddf44e9df4f7b/scala-library-2.11.6.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.apache.commons/commons-lang3/3.4/5fe28b9518e58819180a43a850fbc0dd24b7c050/commons-lang3-3.4.jar], keepAliveMode=SESSION}.
:compilePlayBinaryPlayTwirlTemplates (Thread[Task worker for ':',5,main]) completed. Took 1.965 secs.
:compilePlayBinaryScala (Thread[Task worker for ':',5,main]) started.
This JVM does not support getting OS memory, so no OS memory status updates will be broadcast
Initialized native services in: /home/ymwang/.gradle/native

> Task :compilePlayBinaryScala
Task ':compilePlayBinaryScala' is not up-to-date because:
  Output property 'destinationDir' file /home/ymwang/Downloads/play-2.4/build/playBinary/classes has been removed.
  Output property 'destinationDir' file /home/ymwang/Downloads/play-2.4/build/playBinary/classes/controllers has been removed.
  Output property 'destinationDir' file /home/ymwang/Downloads/play-2.4/build/playBinary/classes/controllers/ReverseApplication.class has been removed.
Starting process 'Gradle Worker Daemon 7'. Working directory: /home/ymwang/.gradle/workers Command: /export/apps/jdk/JDK-1_8_0_121/bin/java -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/ymwang/.gradle/caches/4.9/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Worker Daemon 7'
Successfully started process 'Gradle Worker Daemon 7'
Started Gradle worker daemon (0.372 secs) with fork options DaemonForkOptions{executable=/export/apps/jdk/JDK-1_8_0_121/bin/java, minHeapSize=null, maxHeapSize=null, jvmArgs=[], classpath=[/home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.zinc/zinc/0.3.15/12e1f782684f2702e847faa0994eed4711270687/zinc-0.3.15.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.sbt/incremental-compiler/0.13.15/95e20d00b25a7aae19838009c11578b7e6b258ad/incremental-compiler-0.13.15.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.sbt/compiler-interface/0.13.15/bad996ed4fc3e83b872525e9cd7b80d81b98a324/compiler-interface-0.13.15-sources.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-compiler/2.10.6/9b15174852f5b6bb1edbf303d5722286a0a54011/scala-compiler-2.10.6.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/com.typesafe.sbt/sbt-interface/0.13.15/93fe450d5f5efb111397a34bc1fba0d50368a265/sbt-interface-0.13.15.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-reflect/2.10.6/3259f3df0f166f017ef5b2d385445808398c316c/scala-reflect-2.10.6.jar, /home/ymwang/.gradle/caches/modules-2/files-2.1/org.scala-lang/scala-library/2.10.6/421989aa8f95a05a4f894630aad96b8c7b828732/scala-library-2.10.6.jar], keepAliveMode=SESSION}.
Compiling with Zinc Scala compiler.
Initialized Zinc Scala compiler: 0.554 secs
Prepared Zinc Scala inputs: 0.606 secs
Pruning sources from previous analysis, due to incompatible CompileSetup.
Compiling 7 Scala sources and 1 Java source to /home/ymwang/Downloads/play-2.4/build/playBinary/classes...
Completed Scala compilation: 6.544 secs
:compilePlayBinaryScala (Thread[Task worker for ':',5,main]) completed. Took 7.114 secs.
:processPlayBinaryPlayResources (Thread[Task worker for ':',5,main]) started.

> Task :processPlayBinaryPlayResources
Task ':processPlayBinaryPlayResources' is not up-to-date because:
  Output property 'destinationDir' file /home/ymwang/Downloads/play-2.4/build/playBinary/resources has been removed.
  Output property 'destinationDir' file /home/ymwang/Downloads/play-2.4/build/playBinary/resources/routes has been removed.
  Output property 'destinationDir' file /home/ymwang/Downloads/play-2.4/build/playBinary/resources/application.conf has been removed.
:processPlayBinaryPlayResources (Thread[Task worker for ':',5,main]) completed. Took 0.002 secs.
:createPlayBinaryJar (Thread[Task worker for ':',5,main]) started.

> Task :createPlayBinaryJar
Task ':createPlayBinaryJar' is not up-to-date because:
  Output property 'archivePath' file /home/ymwang/Downloads/play-2.4/build/playBinary/lib/play-2.4.jar has been removed.
:createPlayBinaryJar (Thread[Task worker for ':',5,main]) completed. Took 0.016 secs.
:playBinary (Thread[Task worker for ':',5,main]) started.

> Task :playBinary
Skipping task ':playBinary' as it has no actions.
:playBinary (Thread[Task worker for ':',5,main]) completed. Took 0.0 secs.
:runPlayBinary (Thread[Task worker for ':',5,main]) started.
Initialized native services in: /home/ymwang/.gradle/native
--- (Running the application, auto-reloading is enabled) ---

[info] p.a.l.c.ActorSystemProvider - Starting application default Akka system: application
[info] p.c.s.NettyServer$ - Listening for HTTP on /0:0:0:0:0:0:0:0:9000

> Task :runPlayBinary
Task ':runPlayBinary' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
Starting process 'Gradle Play Worker 8'. Working directory: /home/ymwang/Downloads/play-2.4 Command: /export/apps/jdk/JDK-1_8_0_121/bin/java -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/ymwang/.gradle/caches/4.9/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Play Worker 8'
Successfully started process 'Gradle Play Worker 8'
Running Play App (:runPlayBinary) at http://localhost:9000/
:runPlayBinary (Thread[Task worker for ':',5,main]) completed. Took 1.253 secs.

BUILD SUCCESSFUL in 11s
7 actionable tasks: 7 executed

Reloadable deployment detected. Entering continuous build.
Starting Build
Settings evaluated using settings file '/home/ymwang/Downloads/play-2.4/settings.gradle'.
Projects loaded. Root project using build file '/home/ymwang/Downloads/play-2.4/build.gradle'.
Included projects: [root project 'play-2.4']

> Configure project :
Evaluating root project 'play-2.4' using build file '/home/ymwang/Downloads/play-2.4/build.gradle'.
All projects evaluated.
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
file or directory '/home/ymwang/Downloads/play-2.4/app/assets', not found
Asked to use scala version 2.11 on play < 2.4.6. Will have to use the 2.10 routes compiler
Selected primary task 'runPlayBinary' from project :
Tasks to be executed: [task ':createPlayBinaryAssetsJar', task ':compilePlayBinaryPlayRoutes', task ':compilePlayBinaryPlayTwirlTemplates', task ':compilePlayBinaryScala', task ':processPlayBinaryPlayResources', task ':createPlayBinaryJar', task ':playBinary', task ':runPlayBinary']
:createPlayBinaryAssetsJar (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :createPlayBinaryAssetsJar UP-TO-DATE
Skipping task ':createPlayBinaryAssetsJar' as it is up-to-date.
:createPlayBinaryAssetsJar (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.006 secs.
:compilePlayBinaryPlayRoutes (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :compilePlayBinaryPlayRoutes UP-TO-DATE
Skipping task ':compilePlayBinaryPlayRoutes' as it is up-to-date.
:compilePlayBinaryPlayRoutes (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.002 secs.
:compilePlayBinaryPlayTwirlTemplates (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :compilePlayBinaryPlayTwirlTemplates UP-TO-DATE
Skipping task ':compilePlayBinaryPlayTwirlTemplates' as it is up-to-date.
:compilePlayBinaryPlayTwirlTemplates (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.002 secs.
:compilePlayBinaryScala (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :compilePlayBinaryScala UP-TO-DATE
Skipping task ':compilePlayBinaryScala' as it is up-to-date.
:compilePlayBinaryScala (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.024 secs.
:processPlayBinaryPlayResources (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :processPlayBinaryPlayResources UP-TO-DATE
Skipping task ':processPlayBinaryPlayResources' as it is up-to-date.
:processPlayBinaryPlayResources (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.001 secs.
:createPlayBinaryJar (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :createPlayBinaryJar UP-TO-DATE
Skipping task ':createPlayBinaryJar' as it is up-to-date.
:createPlayBinaryJar (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.003 secs.
:playBinary (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :playBinary UP-TO-DATE
Skipping task ':playBinary' as it has no actions.
:playBinary (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.0 secs.
:runPlayBinary (Thread[Task worker for ':' Thread 2,5,main]) started.

> Task :runPlayBinary
Task ':runPlayBinary' is not up-to-date because:
  Task has not declared any outputs despite executing actions.
:runPlayBinary (Thread[Task worker for ':' Thread 2,5,main]) completed. Took 0.001 secs.

BUILD SUCCESSFUL in 11s
7 actionable tasks: 1 executed, 6 up-to-date

Waiting for changes to input files of tasks... (ctrl-d to exit)
<=============> 100% EXECUTING [17s]
> IDLE

```

**Comment by big-guy** [comment]
Yep. That's what I was seeing locally too.  The second build completes <1 second, but it reports that it's build time is much longer.

I renamed the issue.

**Comment by ymwang1984** [comment]
Thanks @big-guy 

Is there an ETA for Gradle 5.1?

**Comment by ymwang1984** [comment]
For LinkedIn, the issue is not really about the build time display. Multiple teams observed this behavior and would like to see the server up and running "the first time it is loaded". 

For a very large application, even if all steps are UP-TO-DATE the second round, it still adds quite some time to the development cycle. I'm happy to discuss with more details :) Thank you very much.

**Comment by lacasseio** [comment]
This will be fixed in 5.3.

**Comment by ymwang1984** [comment]
@lacasseio  Thanks, is there a nightly that we can check?

**Comment by lacasseio** [comment]
The RC that should be coming tomorrow will have the fix. However, I think the fix was merged back to master last week. You can use the latest nightly.

**Comment by lacasseio** [comment]
@ymwang1984 Just so we can close this issue, the error within the reported time was fixed. However, regarding the duplicated run when the build transition from non-continuous to continuous because of the deployment being generated won't be fixed. If this causes significant pain, you can open another issue explaining how it's affecting you so we can prioritize work around this work.

**Comment by ymwang1984** [comment]
@lacasseio  Thank you for the reply. The reported time is _not_ the issue. 

The issue is truly about the duplicate runs because that's the _actual_ time that we can save for developers. Every day, "runPlayBinary" is invoked more than 4000 times in the company. Removing duplicate runs will have a huge impact.
